mkosi/__init__.py

```diff
@@ -1197,34 +1200,20 @@ def luks_setup_all(
         return
     assert loopdev is not None
 
-    try:
-        root = luks_setup_root(args, loopdev, do_run_build_script)
-        try:
-            home = luks_setup_home(args, loopdev, do_run_build_script)
-            try:
-                srv = luks_setup_srv(args, loopdev, do_run_build_script)
-                try:
-                    var = luks_setup_var(args, loopdev, do_run_build_script)
-                    try:
-                        tmp = luks_setup_tmp(args, loopdev, do_run_build_script)
-
-                        yield LuksSetupOutput(
-                            optional_partition(loopdev, args.root_partno) if root is None else root,
-                            optional_partition(loopdev, args.home_partno) if home is None else home,
-                            optional_partition(loopdev, args.srv_partno) if srv is None else srv,
-                            optional_partition(loopdev, args.var_partno) if var is None else var,
-                            optional_partition(loopdev, args.tmp_partno) if tmp is None else tmp,
-                        )
-                    finally:
-                        luks_close(tmp, "Closing LUKS temporary data partition")
-                finally:
-                    luks_close(var, "Closing LUKS variable data partition")
-            finally:
-                luks_close(srv, "Closing LUKS server data partition")
-        finally:
-            luks_close(home, "Closing LUKS home partition")
-    finally:
-        luks_close(root, "Closing LUKS root partition")
+    with contextlib.ExitStack() as stack:
+        root = stack.enter_context(luks_setup_root(args, loopdev, do_run_build_script))
+        home = stack.enter_context(luks_setup_home(args, loopdev, do_run_build_script))
+        srv = stack.enter_context(luks_setup_srv(args, loopdev, do_run_build_script))
+        var = stack.enter_context(luks_setup_var(args, loopdev, do_run_build_script))
+        tmp = stack.enter_context(luks_setup_tmp(args, loopdev, do_run_build_script))
```

Comment:
We should be able to do this as
```py
with (luks_setup_root(args, loopdev, do_run_build_script) as root,
         luks_setup_home(args, loopdev, do_run_build_script) as home,
         luks_setup_srv(args, loopdev, do_run_build_script) as srv,
         luks_setup_var(args, loopdev, do_run_build_script) as var,
         luks_setup_tmp(args, loopdev, do_run_build_script) as tmp):
             yield …
```

Reply:
Only from python 3.10 onwards.

Reply:
So only three more years, when we target Debian stable. :)